### PR TITLE
Added GroupBy option for actionAdminProductsListingFieldsModifier hook

### PR DESCRIPTION
**Pull request by astterx**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Added group by for sql query to be used with actionAdminProductsListingFieldsModifier hook
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | (optional) If this PR fixes a [Forge](http://forge.prestashop.com/) ticket, please add its complete Forge URL.
| How to test?  | Please indicate how to best verify that this PR is correct.

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8514)

### CODE REVIEW

**Comment by kpodemski** [comment]
Hey @astterx 

you can't just additional parameter inside the function before the previous ones as this is breaking change

File: src/Adapter/Product/AdminProductDataProvider.php [hunk 1]
```diff
@@ -322,13 +325,14 @@ public function getCatalogProductList($offset, $limit, $orderBy, $sortOrder, $po
...
             'sql_order' => &$sqlOrder,
             'sql_limit' => &$sqlLimit,
         ));
 
-        $sql = $this->compileSqlQuery($sqlSelect, $sqlTable, $sqlWhere, $sqlOrder, $sqlLimit);
-        $products = Db::getInstance()->executeS($sql, true, false);
-        $total = Db::getInstance()->executeS('SELECT FOUND_ROWS();', true, false);
+        $sql = $this->compileSqlQuery($sqlSelect, $sqlTable, $sqlWhere, $sqlGroupBy, $sqlOrder, $sqlLimit);
+        $products = \Db::getInstance()->executeS($sql, true, false);
+        $total = \Db::getInstance()->executeS('SELECT FOUND_ROWS();', true, false);
```
Comment (Quetzacoalt91) [inline src/Adapter/Product/AdminProductDataProvider.php]: No need to replace `Db` with `\Db`, as we declared the class at the top of the file with 

```php
use Db;
```

**Comment by Quetzacoalt91** [comment]
Do you have a use case to share with that PR @astterx?

**Comment by astterx** [comment]
For a shop with car parts, I created a module to assign every product to a compatible car. A product can be compatible to multiple cars so using hookActionAdminProductsListingFieldsModifier to retrieve also car model for each product, method getCatalogProductList() will return duplicated products, because now is not possible to group by.

**Comment by astterx** [comment]
Hi @kpodemski,

> Hey @astterx
> 
> you can't just additional parameter inside the function before the previous ones as this is breaking change

Method is used only once, just in location where I modified.

File: src/Adapter/Admin/AbstractAdminQueryBuilder.php [hunk 2]
```diff
@@ -166,6 +167,11 @@ protected function compileSqlQuery(array $select, array $table, array $where = a
             }
         }
 
+        // GROUP BY
+        if (!empty($groupBy)) {
+            $sql[] = 'GROUP BY '. implode(', ', $groupBy) . PHP_EOL;
```
Comment (Quetzacoalt91) [inline src/Adapter/Admin/AbstractAdminQueryBuilder.php]: I feel we may probably handle user data here, and we need to add some sanitization here with

```php
array_map('pSQL', $groupBy);
```

Can you please replace this line with this one:
```php
$sql[] = 'GROUP BY '. implode(', ', array_map('pSQL', $groupBy)) . PHP_EOL;
```

File: src/Adapter/Product/AdminProductDataProvider.php [hunk 3]
```diff
@@ -286,12 +286,15 @@ public function getCatalogProductList($offset, $limit, $orderBy, $sortOrder, $po
             $sqlOrder = array('id_product ASC');
         }
 
+        $sqlGroupBy = [];
```
Comment (Quetzacoalt91) [inline src/Adapter/Product/AdminProductDataProvider.php]: Currently, we initialize arrays with the long syntax form. `[]` should be `array()`

**Review by Quetzacoalt91 — CHANGES_REQUESTED** [review]
Just some changes to bring and we should be good.

About the method signature, I decided to not ask for anything as we deal with a protected one. Overrides are also not available for this part of the code, so broken code should not occur here.

**Review by Quetzacoalt91 — DISMISSED** [review]

**Comment by Quetzacoalt91** [comment]
Thank you @astterx

**Comment by mickaelandrieu** [comment]
@mbadrani what do you think about this improvement? Can we accept it?

Thanks for your contribution @astterx

**Comment by astterx** [comment]
Hi @Quetzacoalt91 ,

Any update?

File: install-dev/data/db_structure.sql [hunk 4]
```diff
@@ -1835,15 +1835,15 @@ CREATE TABLE `PREFIX_search_engine` (
...
+  `id_word` bigint(20) unsigned NOT NULL,
   `weight` smallint(4) unsigned NOT NULL DEFAULT 1,
   PRIMARY KEY (`id_word`, `id_product`),
   KEY `id_product` (`id_product`,`weight`)
 ) ENGINE=ENGINE_TYPE DEFAULT CHARSET=utf8 COLLATION;
 
 /* List of words available for a given shop & lang */
 CREATE TABLE `PREFIX_search_word` (
-  `id_word` int(10) unsigned NOT NULL auto_increment,
+  `id_word` bigint(20) unsigned NOT NULL auto_increment,
```
Comment (Quetzacoalt91) [inline install-dev/data/db_structure.sql]: New changes were brought to this PR but are unrelated to the original subject. :(

**Comment by Quetzacoalt91** [comment]
Hello @astterx, we were waiting for the QA team before further action.
In the meantime, would you mind removing the additional commit not related to the groupBy parameter?

Thank you

**Comment by astterx** [comment]
I removed additional commit.

**Review by Quetzacoalt91 — DISMISSED** [review]
Thank you. Now, this is in the hands of the QA team. ;)

**Comment by marionf** [comment]
Hello @astterx & @Quetzacoalt91 
What are the functional tests to do ?

**Comment by astterx** [comment]
Hi @marionf 
You should do a regression testing on admin product list.

**Comment by astterx** [comment]
Hi. Any update here?

**Comment by marionf** [comment]
ping @Quetzacoalt91 can you review it before I test ?

**Review by Quetzacoalt91 — APPROVED** [review]
Still fine for me.

**Comment by mickaelandrieu** [comment]
Thanks @astterx !

**Pull request merged**